narwhal/consensus/src/consensus.rs

```diff
@@ -31,6 +35,114 @@ pub mod consensus_tests;
+
+        debug!("Reputation scores: {:?}", reputation_scores);
+        debug!("Good nodes: {:?}", good_nodes);
+        debug!("Bad nodes: {:?}", bad_nodes);
+
+        Self {
+            good_nodes,
+            bad_nodes,
+        }
+    }
+
+    /// Checks whether the provided leader is a bad performer and needs to be swapped for the schedule
+    /// with a good performer. If not, then the method returns None. Otherwise the leader to swap with
+    /// is returned instead. The `leader_round` represents the DAG round on which the provided AuthorityIdentifier
+    /// is a leader on and is used as a seed to random function in order to calculate the good node that
+    /// will swap for that round the bad node. We are intentionally not doing weighted randomness as
+    /// we want to give to all the good nodes equal opportunity to get swapped with bad nodes and not
+    /// have one node with enough stake dominating on the final schedule.
+    pub fn swap(
+        &self,
+        leader: &AuthorityIdentifier,
+        leader_round: Round,
+    ) -> Option<AuthorityIdentifier> {
+        if self.bad_nodes.contains(leader) {
+            let mut seed_bytes = [0u8; 32];
+            seed_bytes[32 - 8..].copy_from_slice(&leader_round.to_le_bytes());
+            let mut rng = StdRng::from_seed(seed_bytes);
+
+            let good_node = self
+                .good_nodes
+                .choose(&mut rng)
+                .expect("There should be at least one good node available");
+
+            trace!(
+                "Swapping bad leader {} -> {} for round {}",
+                leader,
+                good_node,
+                leader_round
+            );
+
+            return Some(*good_node);
+        }
+        None
+    }
+
+    /// Retrieves the f by stake nodes with the best scores.
+    fn good_nodes(
+        reputation_scores: &ReputationScores,
+        committee: &Committee,
+    ) -> Vec<AuthorityIdentifier> {
+        let authorities = reputation_scores.authorities_by_score_desc();
+
+        Self::retrieve_first_f_nodes(committee, authorities.into_iter())
+    }
+
+    /// Retrieves the f by stake nodes with the worst scores
+    fn bad_nodes(
+        reputation_scores: &ReputationScores,
+        committee: &Committee,
+    ) -> HashSet<AuthorityIdentifier> {
+        let authorities = reputation_scores.authorities_by_score_desc();
+
+        // we revert the sorted authorities to score ascending so we get the first f low scorers
+        let bad_authorities =
+            Self::retrieve_first_f_nodes(committee, authorities.into_iter().rev());
+        bad_authorities
+            .into_iter()
+            .collect::<HashSet<AuthorityIdentifier>>()
+    }
+
+    // Retrieves the first f by stake nodes provided by the iterator `authorities`. It's the
+    // caller's responsibility to ensure that the elements of the `authorities` input is already
+    // sorted.
+    fn retrieve_first_f_nodes(
```

Comment:
Can we extract `f` out as a const, and play around with it a bit more? Does it prevent leader swap from achieving some properties? With uneven stake distributions, "bad_nodes" with f stake could be more than half of the validators. I need to think about the implication of that. Only swapping a few down or persistently slow nodes should be good enough in practice.

Reply:
We can definitely do, I'll do that as part of upcoming PR. It doesn't affect on achieving properties , when set to lower than `f`. You definitely have a point on the absolute number of nodes - this is something that we also had to deal with the scores on the submission part.

Reply:
Sounds good to adjust the threshold in a future PR. But let's keep f out of the function name?

Reply:
actually the smaller it is the better the liveness gurantees we can get but the higher the risk of a bad leader not proposing on their round